runsc/boot/gofer_conf.go

```diff
@@ -16,52 +16,172 @@ package boot
 	MemoryOverlay
 
-	// SelfOverlay indicates that this gofer mount should be overlaid with an
-	// overlayfs backed by a host file in the mount's source directory.
+	// SelfOverlay indicates that this gofer mount should be overlaid with a
+	// tmpfs backed by a host file in the mount's source directory.
 	SelfOverlay
 
-	// AnonOverlay indicates that this gofer mount should be overlaid with an
-	// overlayfs backed by a host file in an anonymous directory.
+	// AnonOverlay indicates that this gofer mount should be overlaid with a
+	// tmpfs backed by a host file in an anonymous directory.
 	AnonOverlay
 
-	// SelfTmpfs indicates that this gofer mount should be overlaid with a tmpfs
-	// mount backed by a host file in the mount's source directory.
-	SelfTmpfs
+	// UpperMax indicates the number of the valid upper layer types.
+	UpperMax
 )
 
+// String returns a human-readable string representing the upper layer type.
+func (u GoferMountConfUpperType) String() string {
+	switch u {
+	case NoOverlay:
+		return "none"
+	case MemoryOverlay:
+		return "memory"
+	case SelfOverlay:
+		return "self"
+	case AnonOverlay:
+		return "anon"
+	}
+	panic(fmt.Sprintf("Invalid gofer mount config upper layer type: %d", u))
+}
+
+// Set sets the value. Set(String()) should be idempotent.
+func (u *GoferMountConfUpperType) Set(v string) error {
+	switch v {
+	case "none":
+		*u = NoOverlay
+	case "memory":
+		*u = MemoryOverlay
+	case "self":
+		*u = SelfOverlay
+	case "anon":
+		*u = AnonOverlay
+	default:
+		return fmt.Errorf("invalid gofer mount config upper layer type: %s", v)
+	}
+	return nil
+}
+
+// GoferMountConfLowerType describes how lower layer is configured for the gofer mount.
+type GoferMountConfLowerType byte
+
+const (
+	// NoneLower indicates that this gofer mount has no lower layer.
+	NoneLower GoferMountConfLowerType = iota
+
+	// Lisafs indicates that this gofer mount has a LISAFS lower layer.
+	Lisafs
+
+	// Erofs indicates that this gofer mount has an EROFS lower layer.
+	Erofs
+
+	// LowerMax indicates the number of the valid lower layer types.
+	LowerMax
+)
+
+// String returns a human-readable string representing the lower layer type.
+func (l GoferMountConfLowerType) String() string {
+	switch l {
+	case NoneLower:
+		return "none"
+	case Lisafs:
+		return "lisafs"
+	case Erofs:
+		return "erofs"
+	}
+	panic(fmt.Sprintf("Invalid gofer mount config lower layer type: %d", l))
+}
+
+// Set sets the value. Set(String()) should be idempotent.
+func (l *GoferMountConfLowerType) Set(v string) error {
+	switch v {
+	case "none":
+		*l = NoneLower
+	case "lisafs":
+		*l = Lisafs
+	case "erofs":
+		*l = Erofs
+	default:
+		return fmt.Errorf("invalid gofer mount config lower layer type: %s", v)
+	}
+	return nil
+}
+
+// GoferMountConf describes how a gofer mount is configured in the sandbox.
+type GoferMountConf struct {
+	Upper GoferMountConfUpperType `json:"upper"`
+	Lower GoferMountConfLowerType `json:"lower"`
```

Comment:
The json annotations were intentionally added. Because `Container` will be marshalled and saved in the state file.

https://github.com/google/gvisor/blob/8f9211abca9c0564c8c1b1f81d6b28c469b4aba9/runsc/container/container.go#L1044

`GoferMountConfs` is also in `Container`:

https://github.com/google/gvisor/blob/8f9211abca9c0564c8c1b1f81d6b28c469b4aba9/runsc/container/container.go#L140

With the json annotations, the new fields will be more consistent with the existing fields in state file:

```
  "goferMountConfs": [
    {
      "upper": 2,
      "lower": 1
    },
    {
      "upper": 0,
      "lower": 1
    },
    {
      "upper": 0,
      "lower": 1
    },
    {
      "upper": 0,
      "lower": 1
    }
  ]
```

Without the annotations, the first letter would be capitalized:

```
  "goferMountConfs": [
    {
      "Upper": 2,
      "Lower": 1
    },
    {
      "Upper": 0,
      "Lower": 1
    },
    {
      "Upper": 0,
      "Lower": 1
    },
    {
      "Upper": 0,
      "Lower": 1
    }
  ]
```

Reply:
Ah got it.